imap/command/search.go

```diff
@@ -0,0 +1,383 @@
+package command
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ProtonMail/gluon/imap/parser"
+	"github.com/bradenaw/juniper/xslices"
+	"time"
+)
+
+type SearchCommand struct {
+	Charset string
+	Keys    []SearchKey
+}
+
+type SearchKey interface {
+	String() string
+	SanitizedString() string
+}
+
+func (s SearchCommand) String() string {
+	charsetStr := "NONE"
+	if len(s.Charset) != 0 {
+		charsetStr = s.Charset
+	}
+
+	return fmt.Sprintf("SEARCH CHARSET=%v %v", charsetStr, s.Keys)
+}
+
+func (s SearchCommand) SanitizedString() string {
+	charsetStr := "NONE"
+	if len(s.Charset) != 0 {
+		charsetStr = s.Charset
+	}
+
+	return fmt.Sprintf("SEARCH CHARSET=%v %v", charsetStr, xslices.Map(s.Keys, func(v SearchKey) string {
+		return v.SanitizedString()
+	}))
+}
+
+type SearchCommandParser struct {
+	charset        string
+	charsetAllowed bool
```

Comment:
Instead of having this internal parser state, why not just check if the first keyword is CHARSET, if so parse it, and then continue parsing keys?

Reply:
Done